internal/cmd/launcher.go

```diff
@@ -0,0 +1,348 @@
+func NewLauncher(gs *state.GlobalState) *Launcher {
+	defaultExecutor := &currentBinary{}
+	return &Launcher{
+		gs:              gs,
+		provision:       k6buildProvision,
+		commandExecutor: defaultExecutor,
+	}
+}
+
+// Launch executes k6 either by launching a provisioned binary or defaulting to the
+// current binary if this is not necessary.
+// If the fhe fallback is called, it can exit the process so don't assume it will return
+func (l *Launcher) Launch() {
+	// If binary provisioning is not enabled, continue with the regular k6 execution path
+	if !l.gs.Flags.BinaryProvisioning {
+		l.gs.Logger.Debug("Binary Provisioning feature is disabled.")
+		l.commandExecutor.run(l.gs)
+		return
+	}
+
+	l.gs.Logger.
+		Debug("Binary Provisioning feature is enabled.")
+	deps, err := analyze(l.gs, l.gs.CmdArgs[1:])
+	if err != nil {
+		l.gs.Logger.
+			WithError(err).
+			Error("Binary provisioning is enabled but it failed to analyze the dependencies." +
+				" Please, make sure to report this issue by opening a bug report.")
+		l.gs.OSExit(1)
+		return // this is required for testing
+	}
+
+	// if the command does not have dependencies or a custom build
+	if !customBuildRequired(build.Version, deps) {
+		l.gs.Logger.
+			Debug("The current k6 binary already satisfies all the required dependencies," +
+				" it isn't required to provision a new binary.")
+		l.commandExecutor.run(l.gs)
+		return
+	}
+
+	l.gs.Logger.
+		WithField("deps", deps).
+		Info("Binary Provisioning experimental feature is enabled." +
+			" The current k6 binary doesn't satisfy all dependencies, it's required to" +
+			" provision a custom binary.")
+
+	customBinary, err := l.provision(l.gs, deps)
+	if err != nil {
+		l.gs.Logger.
+			WithError(err).
+			Error("Failed to provision a k6 binary with required dependencies." +
+				" Please, make sure to report this issue by opening a bug report.")
+		l.gs.OSExit(1)
+		return
+	}
+
+	customBinary.run(l.gs)
+}
+
+// customBinary runs the requested commands
+// on a different binary on a subprocess passing the original arguments
+type customBinary struct {
+	// path represents the local file path
+	// on the file system of the binary
+	path string
+}
+
+//nolint:forbidigo
+func (b *customBinary) run(gs *state.GlobalState) {
+	cmd := exec.CommandContext(gs.Ctx, b.path, gs.CmdArgs[1:]...) //nolint:gosec
+
+	// we pass os stdout, err, in because passing them from GlobalState changes how
+	// the subprocess detects the type of terminal
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	cmd.Stdin = os.Stdin
+
+	// Copy environment variables to the k6 process and skip binary provisioning feature flag to disable it.
+	// If not disabled, then the executed k6 binary would enter an infinite loop, where it continuously
+	// process the input script, detect dependencies, and retrigger provisioning.
+	env := []string{}
+	for k, v := range gs.Env {
+		if k == "K6_BINARY_PROVISIONING" {
```

Comment:
I'd say I have seen this literal string at least a couple of times along this PR, so I'd prefer if we can move this into a `const`, so prevent mistakes in the future, in case it changes, possible typos, etc.